### PR TITLE
feature: add CORS and Cookie

### PR DESCRIPTION
**Pull request by LBokan**
### Description

Added CORS and Cookie settings

### Type of change

- [x] Feature (introduces new functionality)
- [ ] Bugfix
- [ ] Configuration
- [ ] Refactoring (doesn't change functionality, only the code)
- [ ] Style changes (doesn't affect functionality, only the look of the components)

### PR Checklist

- [x] Synced up with latest from the main branch/Code is linted
- [x] Description of the changes has been added (Screenshots/Video)

> If not explain here

- [x] Commit messages follow [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#summary)
- [x] Console has been checked for warnings

#### Screenshots/Video before change

Add screenshots of the changed components here.

#### Screenshots/Video after change

Add screenshots with the changes applied here.

### CODE REVIEW

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java [hunk 1]
```diff
@@ -75,9 +80,21 @@ public ResponseEntity signup( @RequestBody RegisterRequestEntity request ) {
...
 
+    Cookie[] cookies = httpServletRequest.getCookies();
+    if ( cookies != null ) {
+      for ( Cookie cookie : cookies ) {
+        cookie.setMaxAge( 0 );
+        cookie.setPath( "/" );
+
+        httpServletResponse.addCookie( cookie );
+      }
+    }
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:96]: Why do you need this? What is the intention of copy paste all cookies from client with overriden properties?

Comment (rmaskalenka) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:96]: You can check this cheat sheet for more details on cookie lifecycle https://crypto.stanford.edu/cs142/lectures/cookie.html

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java [hunk 2]
```diff
@@ -88,6 +105,13 @@ public ResponseEntity login( @RequestBody AuthRequestEntity request ) {
     UserEntity user = ( UserEntity ) authentication.getPrincipal();
     String jwtToken = jwtService.generateToken( userEmail );
 
+    Cookie cookie = new Cookie( "token", jwtToken );
+    cookie.setMaxAge( 86400 );
+    cookie.setPath( "/" );
+
+    httpServletResponse.addCookie( cookie );
+    httpServletResponse.setHeader( HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, "Set-Cookie" );
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:113]: You don't need to have `httpServletResponse` to set Cookie. Cookie is one of the headers.
Please refer to `org.springframework.http.ResponseCookie`, `org.springframework.http.HttpHeaders`, `org.springframework.http.ResponseEntity.headers()`

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java [hunk 3]
```diff
@@ -75,9 +80,21 @@ public ResponseEntity signup( @RequestBody RegisterRequestEntity request ) {
   }
 
   @Override
-  public ResponseEntity login( @RequestBody AuthRequestEntity request ) {
+  public ResponseEntity login( @RequestBody AuthRequestEntity request,
+                               HttpServletRequest httpServletRequest,
+                               HttpServletResponse httpServletResponse ) {
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:85]: please don't inject raw `request`\`response` objects. It should be used only as a last resort or when you actually do a lot of operations with input\output properties.

File: backend/src/main/java/com/exadel/carinsurance/model/UserEntity.java [hunk 4]
```diff
@@ -34,6 +35,7 @@ public class UserEntity implements UserDetails {
   private String email;
 
   @Column( name = "password" )
+  @JsonIgnore
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/model/UserEntity.java:38]: it is better to have separate models for DAO and Controller layer. That way it is harder to expose anything unnecessary during development

**Review by rmaskalenka — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java (hunk 1 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:96]: Removed this part of code

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java (hunk 2 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:113]: Done

File: backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java [hunk 5]
```diff
@@ -85,16 +88,28 @@ public ResponseEntity login( @RequestBody AuthRequestEntity request ) {
...
     String jwtToken = jwtService.generateToken( userEmail );
 
+    ResponseCookie cookie = ResponseCookie.from( "token", jwtToken )
+        .path( "/" )
+        .maxAge( 86400 )
+        .build();
+
+    HttpHeaders headers = new HttpHeaders();
+    headers.add( HttpHeaders.SET_COOKIE, cookie.toString() );
+    headers.add( HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, "Set-Cookie" );
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/service/implementation/AuthService.java:102]: [Please ignore] No idea why you want to expose cookies as a part of CORS. This does looks like a major vurability for CSRF attack. But as this is outside of our initial learning scope you can ignore my comment

File: backend/src/main/java/com/exadel/carinsurance/model/auth/UserResponseEntity.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  public static UserResponseEntity fromUserEntity( UserEntity userEntity ) {
+    return UserResponseEntity.builder()
+        .userId( userEntity.getUserId() )
+        .firstName( userEntity.getFirstName() )
+        .lastName( userEntity.getLastName() )
+        .email( userEntity.getEmail() )
+        .role( userEntity.getRole() )
+        .build();
+  }
```
Comment (rmaskalenka) [inline backend/src/main/java/com/exadel/carinsurance/model/auth/UserResponseEntity.java:29]: [Optional] Usually it is not advised to have any logic in your POJO\Entity objects as they should be used only for data transfer. Usually it is advised to provide separate Mapper classes that will handle the mapping for you.

**Review by rmaskalenka — APPROVED** [review]

File: backend/src/main/java/com/exadel/carinsurance/model/auth/UserResponseEntity.java (hunk 6 above)
Comment (LBokan) [reply backend/src/main/java/com/exadel/carinsurance/model/auth/UserResponseEntity.java:29]: Done. Created a mapper class

**Pull request merged**